---
title: "Autism Spectrum Disorder Methylation analysis - comparison of Guthrie card and fresh blood"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/asd_meth

## Introduction

Here I will be running a comparison of differential methylation data from guthrie and fresh blood samples.

Here are the files that I'm using:

* ASD_blood_top_dmps_onADOS_withintw_Nov27_limma.csv

* ASD_guthrie_top_dmps_onADOS_withintw_Nov27_limma.csv


```{r,libs}

suppressPackageStartupMessages({
  library("parallel")
  library("mitch")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  source("https://raw.githubusercontent.com/markziemann/gmea/main/meth_functions.R")
  library("data.table")
  library("kableExtra")
  library("eulerr")
  library("RIdeogram")
  library("GenomicRanges")
  library("tictoc")
})

```


## Load data

```{r,load1}

bl_design <- readRDS(file="bl_design.rds")
bl_mvals <- readRDS(file="bl_mvals.rds")

gu_design <- readRDS(file="gu_design.rds")
gu_mvals <- readRDS(file="gu_mvals.rds")

```

## Probe sets

For each gene, extract out the probes.

```{r,probesets}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])
promoters <- grep("Prom",myann$Regulatory_Feature_Group)
promoters <- myann[promoters,]
gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
sets <- split(rep(names(gp2), lengths(gp2)), unlist(gp2))
summary(unlist(lapply(sets,length)))

```

## Load Limma and RUV data for manhattan plot

First I will generate plots for limmma analysis.

ASD_blood_top_dmps_onADOS_Nov27_withintw_limma_top1k.csv

ASD_guthrie_top_dmps_onADOS_Nov27_withintw_limma_top1k.csv

```{r,manhattan1}

# blood at assessment
top <- read.csv("ASD_blood_top_dmps_onADOS_withintw_Nov27_limma.csv")
nrow(top)
top <- subset(top,P.Value<1e-2)
nrow(top)
-log10(min(top$P.Value))
top$chr <- as.integer(gsub("chr","",top$chr))
top$snp <- paste(top$Name,top$UCSC_RefGene_Name)
top$snp <- sapply(strsplit(top$snp,";"),"[[",1)
up <- subset(top,logFC>0)
dn <- subset(top,logFC<0)

par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 ,main="Blood at assessment limma hypermethylated")
manhattan(x=dn,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 , main="Blood at assessment limma hypomethylated")

pdf("manhat_limma_bl.pdf")
par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 ,main="Blood at assessment limma hypermethylated")
manhattan(x=dn,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 , main="Blood at assessment limma hypomethylated")
dev.off()

# neonatal guthrie card
top <- read.csv("ASD_guthrie_top_dmps_onADOS_withintw_Nov27_limma.csv")
nrow(top)
top <- subset(top,P.Value<1e-2)
nrow(top)
-log10(min(top$P.Value))
top$chr <- as.integer(gsub("chr","",top$chr))
top$snp <- paste(top$Name,top$UCSC_RefGene_Name)
top$snp <- sapply(strsplit(top$snp,";"),"[[",1)
up <- subset(top,logFC>0)
dn <- subset(top,logFC<0)

par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 ,main="Neonatal Guthrie card limma hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 , main="Neonatal Guthrie card limma hypomethylated",
  annotateTop=FALSE)

pdf("manhat_limma_gu.pdf")
par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 ,main="Neonatal Guthrie card limma hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="P.Value",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-04 , main="Neonatal Guthrie card limma hypomethylated",
  annotateTop=FALSE)
dev.off()

```

Now I will do the same for the RUV analysis.

ASD_blood_topruv80pc_1kdmps_onADOS_withintw_Nov29.csv

ASD_guthrie_topruv80%_1kdmps_onADOS_withintw_Nov29.csv


```{r,manhattan2}

# blood at assessment
top <- read.csv("ASD_blood_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
nrow(top)
top <- subset(top,F.p<1e-2)
nrow(top)
-log10(min(top$F.p))
top$chr <- as.integer(gsub("chr","",top$chr))
top$snp <- paste(top$Name,top$UCSC_RefGene_Name)
top$snp <- sapply(strsplit(top$snp,";"),"[[",1)
up <- subset(top,b_X1>0)
dn <- subset(top,b_X1<0)

par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-05 ,main="Blood at assessment RUV hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-05 ,main="Blood at assessment RUV hypomethylated",
  annotateTop=FALSE)

pdf("manhat_ruv_bl.pdf")
par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-05 ,main="Blood at assessment RUV hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 1e-05 ,main="Blood at assessment RUV hypomethylated",
  annotateTop=FALSE)
dev.off()

# Neonatal guthrie card
top <- read.csv("ASD_guthrie_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
nrow(top)
top <- subset(top,F.p<1e-2)
nrow(top)
-log10(min(top$F.p))
top$chr <- as.integer(gsub("chr","",top$chr))
top$snp <- paste(top$Name,top$UCSC_RefGene_Name)
top$snp <- sapply(strsplit(top$snp,";"),"[[",1)
up <- subset(top,b_X1>0)
dn <- subset(top,b_X1<0)

par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 5e-05 ,main="Neonatal Guthrie card RUV hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 5e-05 ,main="Neonatal Guthrie card RUV hypomethylated", 
  annotateTop=FALSE)

pdf("manhat_ruv_gu.pdf")
par(mfrow=c(2,1))
par(mar=c(4,4,3,5))
manhattan(x=up,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 5e-05 ,main="Neonatal Guthrie card RUV hypermethylated",
  annotateTop=FALSE)
manhattan(x=dn,chr="chr",bp="pos",p="F.p",snp="snp",suggestiveline = -log10(1e-04),
  ylim=c(2,6), annotatePval= 5e-05 ,main="Neonatal Guthrie card RUV hypomethylated", 
  annotateTop=FALSE)
dev.off()

```

## Compartment enrichment

Looking for enrichments in different genomic compartments.

```{r,comp1}

compartment_enrichment <- function(dma) {
  up <- subset(dma,b_X1>0 & F.p<1e-4)
  dn <- subset(dma,b_X1<0 & F.p<1e-4)
  all <- table(unique(dma)$Regulatory_Feature_Group)
  up <- table(unique(up)$Regulatory_Feature_Group)
  dn <- table(unique(dn)$Regulatory_Feature_Group)
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(up,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  rownames(xx)[1] <- "Intergenic"
  xx[,1] = NULL
  colnames(xx) <- c("all","up")
  xx[is.na(xx)] <- 0
  head(xx)
  x=xx$up
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$up)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$up)-x[2], sum(xx$all) - sum(xx$up) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  up_comp <- xx

  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(dn,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  rownames(xx)[1] <- "Intergenic"
  xx[,1] = NULL
  colnames(xx) <- c("all","dn")
  xx[is.na(xx)] <- 0
  x=xx$dn
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$dn)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$dn)-x[2], sum(xx$all) - sum(xx$dn) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  dn_comp <- xx
  list("up_comp"=up_comp,"dn_comp"=dn_comp)
}

make_forest_plots <- function(comp) {

  comp_data <- 
    structure(list(
      "mean"  = comp$up_comp$OR , 
      "lower" = comp$up_comp$lowerCI ,
      "upper" = comp$up_comp$upperCI ,
      .Names = c("mean", "lower", "upper"), 
      row.names = c(NA, -11L), 
      class = "data.frame"))

  comp_data <- as.data.frame(comp_data[1:3],row.names = rownames(comp$up_comp) )

  forestplot(comp_data,title = "hypermethylated",
    labeltext = as.list(rownames(comp_data)),
    mean=mean,lower=lower,upper=upper)

comp_data <- 
  structure(list(
    "mean"  = comp$dn_comp$OR , 
    "lower" = comp$dn_comp$lowerCI ,
    "upper" = comp$dn_comp$upperCI ,
    .Names = c("mean", "lower", "upper"), 
    row.names = c(NA, -11L), 
    class = "data.frame"))

comp_data <- as.data.frame(comp_data[1:3],row.names = rownames(comp$dn_comp) )

  forestplot(comp_data,title = "hypomethylated",
    labeltext = as.list(rownames(comp_data)),
    mean=mean,lower=lower,upper=upper)

}

par(mfrow=c(2,1))
# guthrie
dma1 <- read.csv("ASD_guthrie_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
comp <- compartment_enrichment(dma1)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

# blood
dma2 <- read.csv("ASD_blood_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
comp <- compartment_enrichment(dma2)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

```

## Compartments top 1000

```{r,comp2}

compartment_enrichment2 <- function(dma) {
  all <- table(unique(dma)$Regulatory_Feature_Group)
  dma <- head(dma,1000)
  up <- subset(dma,b_X1>0 )
  dn <- subset(dma,b_X1<0 )
  up <- table(unique(up)$Regulatory_Feature_Group)
  dn <- table(unique(dn)$Regulatory_Feature_Group)
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(up,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  rownames(xx)[1] <- "Intergenic"
  xx[,1] = NULL
  colnames(xx) <- c("all","up")
  xx[is.na(xx)] <- 0
  head(xx)
  x=xx$up
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$up)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$up)-x[2], sum(xx$all) - sum(xx$up) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  up_comp <- xx

  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(dn,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  rownames(xx)[1] <- "Intergenic"
  xx[,1] = NULL
  colnames(xx) <- c("all","dn")
  xx[is.na(xx)] <- 0
  x=xx$dn
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$dn)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$dn)-x[2], sum(xx$all) - sum(xx$dn) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  dn_comp <- xx
  list("up_comp"=up_comp,"dn_comp"=dn_comp)
}

par(mfrow=c(2,1))
# guthrie
dma1 <- read.csv("ASD_guthrie_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
comp <- compartment_enrichment2(dma1)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

# blood
dma2 <- read.csv("ASD_blood_topruv80pc_dmps_onADOS_withintw_Nov29.csv")
comp <- compartment_enrichment2(dma2)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

```

## Looking for enrichments in proximity to CGI contexts

```{r,cgi1}

cgi_enrichment <- function(dma) {
  dma$Relation_to_Island <- gsub("N_","",dma$Relation_to_Island)
  dma$Relation_to_Island <- gsub("S_","",dma$Relation_to_Island)
  all <- table(unique(dma)$Relation_to_Island)
  up <- subset(dma,b_X1>0 & F.p < 1e-4)
  dn <- subset(dma,b_X1<0 & F.p < 1e-4)
  up <- table(unique(up)$Relation_to_Island)
  dn <- table(unique(dn)$Relation_to_Island)
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(up,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  xx[,1] = NULL
  colnames(xx) <- c("all","up")
  xx[is.na(xx)] <- 0
  head(xx)
  x=xx$up
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$up)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$up)-x[2], sum(xx$all) - sum(xx$up) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  up_comp <- xx
  
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(dn,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  xx[,1] = NULL
  colnames(xx) <- c("all","dn")
  xx[is.na(xx)] <- 0
  x=xx$dn
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$dn)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$dn)-x[2], sum(xx$all) - sum(xx$dn) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  dn_comp <- xx
  list("up_comp"=up_comp,"dn_comp"=dn_comp)
}

par(mfrow=c(2,1))
# guthrie
comp <- cgi_enrichment(dma1)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

# blood
comp <- cgi_enrichment(dma2)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

```

## CGI enrichments in top 1000 probes


```{r,cgi2}

cgi_enrichment2 <- function(dma) {
  dma$Relation_to_Island <- gsub("N_","",dma$Relation_to_Island)
  dma$Relation_to_Island <- gsub("S_","",dma$Relation_to_Island)
  all <- table(unique(dma)$Relation_to_Island)
  dma <- head(dma,1000)
  up <- subset(dma,b_X1>0 )
  dn <- subset(dma,b_X1<0 )
  up <- table(unique(up)$Relation_to_Island)
  dn <- table(unique(dn)$Relation_to_Island)
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(up,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  xx[,1] = NULL
  colnames(xx) <- c("all","up")
  xx[is.na(xx)] <- 0
  head(xx)
  x=xx$up
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$up)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$up)-x[2], sum(xx$all) - sum(xx$up) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  up_comp <- xx
  
  xx=NULL
  xx <- merge(as.data.frame(all, row.names = 1),as.data.frame(dn,row.names = 1),by=0, all = TRUE)
  rownames(xx) <- xx[,1]
  xx[,1] = NULL
  colnames(xx) <- c("all","dn")
  xx[is.na(xx)] <- 0
  x=xx$dn
  m=xx$all
  n=sum(xx$all)-xx$all
  k=sum(xx$dn)
  xl <- apply(xx,1,function(x) {
    mat <- matrix(c(x[2],x[1]-x[2], sum(xx$dn)-x[2], sum(xx$all) - sum(xx$dn) -x [1] + x[2] ),2,2)
    mat
    fisher.test(mat) 
  })
  xx$OR <- unname(unlist(lapply(X=xl, FUN = function(x) {x$estimate})))
  xx$fisherPval <- unname(unlist(lapply(X=xl, FUN = function(x) {x$p.value})))
  xx$lowerCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[1]]})))
  xx$upperCI <- unname(unlist(lapply(X=xl, FUN = function(x) {x$conf.int[[2]]})))
  dn_comp <- xx
  list("up_comp"=up_comp,"dn_comp"=dn_comp)
}

par(mfrow=c(2,1))
# guthrie
comp <- cgi_enrichment2(dma1)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

# blood
comp <- cgi_enrichment2(dma2)
comp <- lapply(comp,function(x) {subset(x,OR!=0)} )
comp
make_forest_plots(comp)

remove(dma1)
remove(dma2)

```

## GMEA functions

This is an enrichment technique.

```{r,func}

# get the median, mean and 1-sample t-test result for each gene
pmea <- function(mval,design,sets,cores=2) {
  fit <- lmFit(mval, design)
  fit <- eBayes(fit)
  top <- topTable(fit,coef=ncol(design),num=Inf, sort.by = "P")
  l <- mclapply(seq(1,length(sets)), function(i) {
    g <- names(sets[i])
    tstats <- top[rownames(top) %in% sets[[i]],"t"]
    myn <- length(tstats)
    mymean <- mean(tstats)
    mymedian <- median(tstats)
    if ( length(tstats) < 2 ) {
      pval=1
    } else {
      wtselfcont <- t.test(tstats)
      pval=wtselfcont$p.value
    }
    c("gene"=g,"nprobes"=myn,"mean"=mymean,"median"=mymedian,
      "P.Value"=pval)
  } , mc.cores=cores)

  df <- do.call(rbind, l)
  rownames(df) <- df[,1]
  df <- df[,-1]
  tmp <- apply(df,2,as.numeric)
  rownames(tmp) <- rownames(df)
  df <- as.data.frame(tmp)
  df$sig <- -log10(df[,4])
  df <- df[order(-df$sig),]
  df$FDR <- p.adjust(df$P.Value)
  out <- list("df"=df,"toptable"=top)
  return(out)
}
# pmea_res <- pmea(mval=mval,design=design,sets=head(sets,20),cores=detectCores()/2)

# Run the fry test for each gene. This is a more conservative test.
run_fry <- function(mval,design,sets,cores=2) {
  split_sets <- split(sets, ceiling(seq_along(sets)/200))
  fry_l <- mclapply(split_sets,function(l) {
    fry(y=mval, index = l, design = design,
      contrast = ncol(design) )
  } , mc.cores=cores )
  fry_res <- do.call(rbind,fry_l)
  rownames(fry_res) <- sub("\\.","@",rownames(fry_res))
  rownames(fry_res) <- sapply(strsplit(rownames(fry_res),"@"),"[[",2)
  fry_res[is.na(fry_res$PValue),"PValue"] <- 1
  fry_res <- fry_res[order(fry_res$PValue),]
  fry_res$FDR <- p.adjust(fry_res$PValue,method="fdr")
  return(fry_res)
}
#fry_res <- run_fry(mval=mval,design=design,sets=sets,cores=cores)

# main function to perform 1-sample t-test and fry test and merge the results.
main <- function(mval,design,sets,cores=2){
  pmea_res <- pmea(mval=mval,design=design,sets=sets,cores=cores)
  pmea_df <- pmea_res[[1]]
  limma_df <- pmea_res[[2]]
  fry_res <- run_fry(mval=mval,design=design,sets=sets,cores=cores)
  m <- merge(pmea_df,fry_res,by=0)
  rownames(m) <- m$Row.names
  m$Row.names = NULL
  m <- m[,c("nprobes","median","PValue","FDR.y")]
  colnames(m) <- c("nprobes","median","PValue","FDR")
  m <- m[order(m$PValue),]
  out <- list("res"=m,"limma_df"=limma_df)
  return(out)
}
#res <- main(mval,design,sets,cores=detectCores()/2)

probe_bias <- function(res) {
  res$sig <- -log10(res$PValue)
  sig <- subset(res,FDR < 0.05)
  plot(res$nprobes,res$sig,log="x",
    pch=19,cex=0.6,
    xlab="no. probes",ylab="-log10(p-value)")
  points(sig$nprobes,sig$sig,col="red",pch=19,cex=0.62)
  SIG = nrow(sig)
  TOT = nrow(res)
  HEADER <- paste(TOT, "genes in total.", SIG, "with FDR<0.05.")
  mtext(HEADER)
}

volcano_plot <- function(res) {
  res$sig <- -log10(res$PValue)
  sig <- subset(res,FDR < 0.05)
  plot(res$median,res$sig,
    pch=19,cex=0.6,
    xlab="median t-statistic",ylab="-log10(p-value)")
  points(sig$median,sig$sig,col="red",pch=19,cex=0.62)
  SIG = nrow(sig)
  UP = nrow(subset(sig,median>0))
  DN = nrow(subset(sig,median<0))
  TOT = nrow(res)
  HEADER <- paste(TOT, "genes in total.", SIG, "with FDR<0.05;",DN,"down,",UP,"up")
  mtext(HEADER)
}

gmea_boxplot <- function(res,sets,n=50) {
  df <- res[[1]]
  limma_df <- res[[2]]
  # smallest pval
  par(mfrow=c(1,2))
  gs <- head(rownames(df),n)
  mysets <- sets[names(sets) %in% gs]
  tstats <- lapply(mysets, function(set) {
    limma_df[rownames(limma_df) %in% set,"t"]
  })
  tstats <- tstats[order(unlist(lapply(tstats,median)))]
  boxplot(tstats,horizontal=TRUE,las=1,
    main="smallest p-val",cex.axis=0.6,
    xlab="t-statistic")
  grid()
  # biggest effect size (median)
  sig <- subset(df,FDR < 0.05)
  gs <- head(rownames(sig[order(-abs(sig$median)),]),n)
  if ( length(gs) >2 ) {
    tstats <- lapply(gs, function(g) {
      df[which(df$genes==g),"tvals"]
    })
    names(tstats) <- gs
    tstats <- tstats[order(unlist(lapply(tstats,median)))]
    boxplot(tstats,horizontal=TRUE,las=1,
      main="biggest effect size(median)",cex.axis=0.6,
      xlab="t-statistic")
    grid()
  } else {
    plot(1)
    mtext("too few significant genes found")
  }
    par(mfrow=c(1,1))
}

```

## Session information

For reproducibility

```{r,session}

sessionInfo()

```
