---
title: "Correlates of ADOS and methylation - enrichment analysis"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/asd_meth

## Introduction

Here I will be running a comparison of differential methylation data from guthrie and fresh blood samples.

Here are the files that I'm using:

* limma_blood_ADOS.csv

* limma_guthrie_ADOS.csv

* limma_buccal_ADOS.csv

```{r,libs}

suppressPackageStartupMessages({
  library("limma")
  library("parallel")
  library("mitch")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  source("meth_functions.R")
  library("data.table")
  library("kableExtra")
  library("eulerr")
  library("GenomicRanges")
  library("RhpcBLASctl")
  data("dualmap850kEID")
})

source("meth_functions.R")

RhpcBLASctl::blas_set_num_threads(1)

```

## Load data

```{r,load1}

bl_mvals <- readRDS(file="bl_mvals.rds")
gu_mvals <- readRDS(file="gu_mvals.rds")
bu_mvals <- readRDS(file="buccal_mvals.rds")

bl_design <- readRDS(file="bl_design_ados.rds")
gu_design <- readRDS(file="gu_design_ados.rds")
bu_design <- readRDS(file="buccal_design_ados.rds")

bl_lim <- read.csv("limma_blood_ADOS.csv")
gu_lim <- read.csv("limma_guthrie_ADOS.csv")
bu_lim <- read.csv("limma_buccal_ADOS.csv")


```

## Probe sets

For each gene, extract out the probes.

```{r,probesets}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])
promoters <- grep("Prom",myann$Regulatory_Feature_Group)
promoters <- myann[promoters,]
gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
sets <- split(rep(names(gp2), lengths(gp2)), unlist(gp2))
summary(unlist(lapply(sets,length)))

```

## Gene methylation enrichment for blood

For each gene, calculate the mean probe t-statistic.

```{r,bl1}

bl <- pmea(mval=bl_mvals,design=bl_design,sets,cores=10)
str(bl)
blm <- bl$df[,"mean",drop=FALSE]
blm <- blm[!is.na(blm$mean),,drop=FALSE] # remove NA vals
head(blm)
nrow(blm)

par(mfrow=c(2,1))
hist(bl$toptable$t,xlab="probe t-stat",main="blood at assessment", breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
hist(bl$df$mean,xlab="gene mean t-stat",main="blood at assessment", breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)

pdf("hist_bl_ados.pdf")
par(mfrow=c(2,1))
hist(bl$toptable$t,xlab="probe t-stat",main="blood at assessment",breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
hist(bl$df$mean,xlab="gene mean t-stat",main="blood at assessment", breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
dev.off()
par(mfrow=c(1,1))

gmea_boxplot(bl,sets)

```

## Gene methylation enrichment for guthrie cards

```{r,gu1}

gu <- pmea(mval=gu_mvals,design=gu_design,sets,cores=10)
str(gu)
gum <- gu$df[,"mean",drop=FALSE]
gum <- gum[!is.na(gum$mean),,drop=FALSE] # remove NA vals
head(gum)
nrow(gum)

par(mfrow=c(2,1))
hist(gu$toptable$t,xlab="probe t-stat",main="neonatal Guthrie card",breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
hist(gu$df$mean,xlab="gene mean t-stat",main="neonatal Guthrie card",breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)

pdf("hist_gu_ados.pdf")
par(mfrow=c(2,1))
hist(gu$toptable$t,xlab="probe t-stat",main="neonatal Guthrie card", breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
hist(gu$df$mean,xlab="gene mean t-stat",main="neonatal Guthrie card", breaks=30) ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
dev.off()
par(mfrow=c(1,1))

gmea_boxplot(gu,sets)

```

## Gene methylation enrichment for buccal samples

```{r,buccal1}

bu <- pmea(mval=bu_mvals,design=bu_design,sets,cores=10)
str(bu)
bum <- bu$df[,"mean",drop=FALSE]
bum <- bum[!is.na(bum$mean),,drop=FALSE] # remove NA vals
head(bum)
nrow(bum)

par(mfrow=c(2,1))
hist(bu$toptable$t,xlab="probe t-stat",main="buccal swab") ; grid() ; abline(v=0,lty=1,lwd=3)
hist(bu$df$mean,xlab="gene mean t-stat",main="buccal swab") ; grid() ; abline(v=0,lty=1,lwd=3)

pdf("hist_bu_ados.pdf")
par(mfrow=c(2,1))
hist(bu$toptable$t,xlab="probe t-stat",main="buccal swab") ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
hist(bu$df$mean,xlab="gene mean t-stat",main="buccal swab") ; grid() ; abline(v=0,lty=1,lwd=3)
plot(1)
dev.off()
par(mfrow=c(1,1))

gmea_boxplot(bu,sets)

```

## Comparison of blood and guthrie card

Now compare gene methylation scores for blood and guthrie cards.
The mean limma t-statistic is used for a scatterplot, then rank-rank plot.

```{r,compare1}

m1 <- merge(blm,gum,by=0)
rownames(m1) <- m1$Row.names
m1$Row.names=NULL
colnames(m1) <- c("bl","gu")

plot(m1$bl,m1$gu,xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9)
grid()
abline(v=0,lty=1,lwd=3,col="black") ; abline(h=0,lty=1,lwd=3,col="black")
mtext("aggregated t-statistic for each gene (mean)")
rm1 <- apply(m1,2,rank)
mydiff <- apply(m1,2,function(x) { length(which(x<0)) } )
rm2 <- rm1
rm2[,1] <- rm2[,1] - mydiff[1]
rm2[,2] <- rm2[,2] - mydiff[2]
rnk <- rm2

palette <- colorRampPalette(c("white", "yellow", "orange", "red","darkred", "black"))
xmin = min(rnk[, 1])
xmax = max(rnk[, 1])
ymin = min(rnk[, 2])
ymax = max(rnk[, 2])
k <- MASS::kde2d(rnk[, 1], rnk[, 2])
X_AXIS = "blood at assessment"
Y_AXIS = "neonatal Guthrie card"

filled.contour(k, xlim = c(xmin, xmax), ylim = c(ymin, ymax),
  color.palette = palette, plot.title = {
  abline(v = 0, h = 0, lty = 2, lwd = 2, col = "blue")
  title(main="gene rank",xlab = X_AXIS, ylab = Y_AXIS)
})

pdf("genedens_ados.pdf")
par(mfrow=c(2,2))
plot(1)
plot(1)

plot(m1$bl,m1$gu,xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9)
grid()
abline(v=0, h=0, lty=2, lwd=2, col="blue")
mtext("aggregated t-statistic for each gene (mean)")

filled.contour(k, xlim = c(xmin, xmax), ylim = c(ymin, ymax),
  color.palette = palette, plot.title = {
  abline(v = 0, h = 0, lty = 2, lwd = 2, col = "blue")
  title(main = "gene rank", xlab = X_AXIS, ylab = Y_AXIS)
})

dev.off()

```

Now we need a table of genes with top scores.

```{r,topgenes1}

m3 <- m1
m3$med <- apply(rnk,1,median)
mg <- merge(gu$df,bl$df,by=0)
rownames(mg) <- mg$Row.names
mg <- mg[rownames(mg) %in% rownames(m3),]
dim(mg)
mg$med <- m3$med
mg <- mg[order(mg$med),]
mg <-subset(mg,nprobes.x>=5)
mg <-subset(mg,nprobes.x>=5)
dim(mg)
mg$Row.names = mg$sig.x = mg$sig.y = mg$FDR.x = mg$FDR.y = mg$median.x = mg$median.y = mg$nprobes.y = NULL

head(mg,20) %>%
  kbl(caption="Genes with coordinated hypomethylation in neonatal Guthrie card and blood at assessment") %>%
  kable_paper("hover", full_width = F)

```

## Gene set level enrichment

```{r,genesets1}

genesets <- gmt_import("ReactomePathways.gmt")

sfari <- readLines("sfari.txt")
lsfari <- list(sfari)
names(lsfari) <- "sfari"

cores = 8

```

## Mitch enrichment analysis

Here I'm using the mean t-statistic for downstream enrichment.
These have already been saved to objects called `blm`, `gum` and `bum`.
Pathways are from REACTOME and analysis using mitch.

```{r,mitch_direction}

mitch_bl <- mitch_calc(blm,genesets,priority="effect")

sig <- subset(mitch_bl$enrichment_result,`p.adjustANOVA`<0.01)

head(sig[order(-abs(sig$s.dist)),],30) %>%
  kbl(caption="BLOOD: Top effect size pathways found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

mitch_gu <- mitch_calc(gum,genesets,priority="effect")

sig <- subset(mitch_gu$enrichment_result,`p.adjustANOVA`<0.01)

head(sig[order(-abs(sig$s.dist)),],30) %>%
  kbl(caption="GUTHRIE: Top effect size pathways found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

mitch_bu <- mitch_calc(bum,genesets,priority="effect")

sig <- subset(mitch_bu$enrichment_result,`p.adjustANOVA`<0.01)

head(sig[order(-abs(sig$s.dist)),],30) %>%
  kbl(caption="BUCCAL: Top effect size pathways found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

mitch_plots(res=mitch_bu,outfile="buccal_mitch_ados.pdf")
mitch_report(res=mitch_bu,outfile="buccal_mitch_ados.html")

subset(mitch_bu$enrichment_result,p.adjustANOVA<0.01 & s.dist<0) %>%
  kbl(caption="Reactome pathways with an inverse methylation association with ADOS in buccal swab samples.") %>%
  kable_paper("hover", full_width = F)

```

Mitch joint analysis of blood and guthrie card.
As using the median value is apparently the most sensitive, we'll go with that.

```{r,mitch_joint}

m <- merge(blm,gum,by=0)
rownames(m) <- m$Row.names
m$Row.names = NULL
colnames(m) <- c("bl","gu")
m[is.na(m)] <- 0
res <- mitch_calc(m, genesets, priority="effect")
mitch_report(res,"blgu_mitch.html",overwrite=TRUE)
mitch_plots(res=res,outfile="blgu_mitch_ados.pdf")
sig <- subset(res$enrichment_result,p.adjustMANOVA<0.01)

head(sig[order(-abs(sig$s.dist)),],30) %>%
  kbl(caption="Top pathways found with mitch (effect size after 1% FDR filtering)") %>%
  kable_paper("hover", full_width = F)

```

Now look at the pathways with hypomethylation.

```{r,hypo}

hypo <- subset(res$enrichment_result,p.adjustMANOVA<0.01 & s.bl <0 & s.gu < 0 )
nrow(hypo)

hypo <- hypo[nrow(hypo):1,]

par(mar=c(4.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set,
  col=gray.colors(2),
  xlab="enrichment score")

legend("topright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"),
  fill=gray.colors(2), horiz=FALSE, cex=1)

pdf("adosbarplot.pdf")

par(mar=c(4.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set, 
  col=gray.colors(2),
  xlab="enrichment score")

legend("topright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"), 
  fill=gray.colors(2), horiz=FALSE, cex=1)

dev.off()


mar = c(5.1, 4.1, 4.1, 2.1)


```

Now look at metatropic glutamate/pheromone receptors.

```{r,glutamate}

phero <- genesets[grep("pheromone",names(genesets))]
phero
rnk2 <- rnk[which(rownames(rnk) %in% phero[[1]]),]

plot(rnk2,main=names(phero),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

pdf("phero.pdf")

plot(rnk2,main=names(phero),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

dev.off()

```

Make a heatmap of top TAS2R gene probes.

```{r,tar2r}

colcolfunc <- colorRampPalette(c("yellow", "orange"))

colcols <- colcolfunc(20)[bl_design[,"ADOS"]]

phero_probes <- unlist(sets[which(names(sets) %in% phero[[1]])])

phero_mvals <- bl_mvals[rownames(bl_mvals) %in% phero_probes,]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(phero_mvals,trace="none",col=colfunc(25), margins = c(5,8), scale="row",
  main="Class C/3 (Metabotropic glutamate/pheromone receptors) probes",
  ColSideColors = colcols )


```

Now let's take a look at the SFARI gene list.

```{r,mitch_SFARI}

str(lsfari)

sfarires <- mitch_calc(m1, lsfari, priority="effect")
t(sfarires$enrichment_result)

lrand <-lapply(1:1034,function(i) { set.seed(i*1000) ; sample(rownames(m1),1000) })
names(lrand) <- paste("random",1:1034)
lsets <- c(lsfari,lrand)

sfarires2 <- mitch_calc(m1, lsets, priority="effect",cores=16,resrows=10)

head(sfarires2$enrichment_result) %>%
  kbl(caption = "Enrichment analysis of SFARI genes") %>%
  kable_paper("hover", full_width = F)

mitch_plots(res=sfarires2,outfile="sfari_mitch_ados.pdf")

mitch_report(res=sfarires2,outfile="sfari_mitch_ados.html",overwrite=TRUE)

```

## Session information

For reproducibility

```{r,session}

sessionInfo()

```
