---
title: "Correlates of ADOS and methylation - enrichment analysis"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/asd_meth

## Introduction

Here I will be running a comparison of differential methylation data from guthrie and fresh blood samples.

Here are the files that I'm using:

* limma_blood_ADOS.csv

* limma_guthrie_ADOS.csv

* limma_buccal_ADOS.csv

```{r,libs}

suppressPackageStartupMessages({
  library("limma")
  library("parallel")
  library("mitch")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  source("meth_functions.R")
  library("data.table")
  library("kableExtra")
  library("eulerr")
  library("GenomicRanges")
  library("HGNChelper")
  library("gplots")

  #data("dualmap850kEID")
})

#source("meth_functions.R")

```

## Probe sets

Make a table that matches probes to genes, keeping in mind that old gene names need to be updated.

```{r,probesets}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","UCSC_RefGene_Group","Islands_Name","Relation_to_Island")])
gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
gp2 <- lapply(gp2,unique)
gt <- stack(gp2)
colnames(gt) <- c("gene","probe")
gt$probe <- as.character(gt$probe)
dim(gt)


#new.hgnc.table <- getCurrentHumanMap()
new.hgnc.table <- readRDS("new.hgnc.table.rds")
fix <- checkGeneSymbols(gt$gene,map=new.hgnc.table)
fix2 <- fix[which(fix$x != fix$Suggested.Symbol),]
gt$gene <- fix$Suggested.Symbol
head(gt)

length(unique(fix2$x))

length(unique(gt$gene))

length(unique(fix2$x)) / length(unique(gt$gene))

```

## Load gene sets

Reactome version Sept 2023.

```{r,genesets}

msigdb <- gmt_import("msigdb.v2023.2.Hs.symbols.gmt")

gobp <- msigdb[grep("^GOBP_",names(msigdb))]
names(gobp) <- gsub("GOBP_","",names(gobp))
names(gobp) <- gsub("_"," ",names(gobp))

reactome <- gmt_import("ReactomePathways.gmt")

```

## Load limma data

```{r,load}

gu_ados <- read.csv("limma_guthrie_ADOS.csv",row.names=1)
gu_diag <- read.csv("limma_guthrie_diagnosis.csv",row.names=1)
gu_iiq <- read.csv("limma_guthrie_iIQ.csv",row.names=1)
gu_ilan <- read.csv("limma_guthrie_ilanguage.csv",row.names=1)
gu_mot <- read.csv("limma_guthrie_motor.csv",row.names=1)

bl_ados <- read.csv("limma_blood_ADOS.csv",row.names=1)
bl_diag <- read.csv("limma_blood_diagnosis.csv",row.names=1)
bl_iiq <- read.csv("limma_blood_iIQ.csv",row.names=1)
bl_ilan <- read.csv("limma_blood_ilanguage.csv",row.names=1)
bl_mot <- read.csv("limma_blood_motor.csv",row.names=1)

buc_ados <- read.csv("limma_buccal_ADOS.csv",row.names=1)
buc_diag <- read.csv("limma_buccal_diagnosis.csv",row.names=1)
buc_iiq <- read.csv("limma_buccal_iIQ.csv",row.names=1)
buc_ilan <- read.csv("limma_buccal_ilanguage.csv",row.names=1)
buc_mot <- read.csv("limma_buccal_motor.csv",row.names=1)

```

## Mitch import

```{r,mitchimport}

m_gu_ados <- mitch_import(x=gu_ados, DEtype="limma", geneTable=gt )
m_gu_diag <- mitch_import(x=gu_diag, DEtype="limma", geneTable=gt )
m_gu_iiq <- mitch_import(x=gu_iiq, DEtype="limma", geneTable=gt )
m_gu_ilan <- mitch_import(x=gu_ilan, DEtype="limma", geneTable=gt )
m_gu_mot <- mitch_import(x=gu_mot, DEtype="limma", geneTable=gt )

m_bl_ados <- mitch_import(x=bl_ados, DEtype="limma", geneTable=gt )
m_bl_diag <- mitch_import(x=bl_diag, DEtype="limma", geneTable=gt )
m_bl_iiq <- mitch_import(x=bl_iiq, DEtype="limma", geneTable=gt )
m_bl_ilan <- mitch_import(x=bl_ilan, DEtype="limma", geneTable=gt )
m_bl_mot <- mitch_import(x=bl_mot, DEtype="limma", geneTable=gt )

m_buc_ados <- mitch_import(x=buc_ados, DEtype="limma", geneTable=gt )
m_buc_diag <- mitch_import(x=buc_diag, DEtype="limma", geneTable=gt )
m_buc_iiq <- mitch_import(x=buc_iiq, DEtype="limma", geneTable=gt )
m_buc_ilan <- mitch_import(x=buc_ilan, DEtype="limma", geneTable=gt )
m_buc_mot <- mitch_import(x=buc_mot, DEtype="limma", geneTable=gt )

```

## Histograms

```{r,hist1,fig.height=4,fig.width=7}

hist(gu_ados$t,xlab="probe t",main="Guthrie ADOS",breaks=40)
hist(gu_diag$t,xlab="probe t",main="Guthrie diagnosis",breaks=40)
hist(gu_iiq$t,xlab="probe t",main="Guthrie inverse IQ",breaks=40)
hist(gu_ilan$t,xlab="probe t",main="Guthrie inverse language",breaks=40)
hist(gu_mot$t,xlab="probe t",main="Guthrie motor skill",breaks=40)

hist(bl_ados$t,xlab="probe t",main="Blood at assessment ADOS",breaks=40)
hist(bl_diag$t,xlab="probe t",main="Blood at assessment diagnosis",breaks=40)
hist(bl_iiq$t,xlab="probe t",main="Blood at assessment inverse IQ",breaks=40)
hist(bl_ilan$t,xlab="probe t",main="Blood at assessment language",breaks=40)
hist(bl_mot$t,xlab="probe t",main="Blood at assessment motor skill",breaks=40)

hist(buc_ados$t,xlab="probe t",main="Buccal ADOS",breaks=40)
hist(buc_diag$t,xlab="probe t",main="Buccal diagnosis",breaks=40)
hist(buc_iiq$t,xlab="probe t",main="Buccal inverse IQ",breaks=40)
hist(buc_ilan$t,xlab="probe t",main="Buccal language",breaks=40)
hist(buc_mot$t,xlab="probe t",main="Buccal motor skill",breaks=40)

hist(m_gu_ados[,1],breaks=40,xlab="t (gene level)",main="Guthrie ADOS")
hist(m_gu_diag[,1],breaks=40,xlab="t (gene level)",main="Guthrie diagnosis")
hist(m_gu_iiq[,1],breaks=40,xlab="t (gene level)",main="Guthrie inverse IQ")
hist(m_gu_ilan[,1],breaks=40,xlab="t (gene level)",main="Guthrie inverse language")
hist(m_gu_mot[,1],breaks=40,xlab="t (gene level)",main="Guthrie motor skills")

hist(m_bl_ados[,1],breaks=40,xlab="t (gene level)",main="Blood at assessment ADOS")
hist(m_bl_diag[,1],breaks=40,xlab="t (gene level)",main="Blood at assessment diagnosis")
hist(m_bl_iiq[,1],breaks=40,xlab="t (gene level)",main="Blood at assessment inverse IQ")
hist(m_bl_ilan[,1],breaks=40,xlab="t (gene level)",main="Blood at assessment inverse language")
hist(m_bl_mot[,1],breaks=40,xlab="t (gene level)",main="Blood at assessment motor skills")

hist(m_buc_ados[,1],breaks=40,xlab="t (gene level)",main="Buccal ADOS")
hist(m_buc_diag[,1],breaks=40,xlab="t (gene level)",main="Buccal diagnosis")
hist(m_buc_iiq[,1],breaks=40,xlab="t (gene level)",main="Buccal inverse IQ")
hist(m_buc_ilan[,1],breaks=40,xlab="t (gene level)",main="Buccal inverse language")
hist(m_buc_mot[,1],breaks=40,xlab="t (gene level)",main="Buccal motor skills")

```

## Scatterplots

```{r,scatter}

par( mar = c(5.1, 4.1, 4.1, 2.1) )

plot(m_bl_ados[,1],m_gu_ados[,1],xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9, main="ADOS")

plot(m_bl_diag[,1],m_gu_diag[,1],xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9, main="Diagnosis")

plot(m_bl_iiq[,1],m_gu_iiq[,1],xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9, main="inverse IQ")

plot(m_bl_ilan[,1],m_gu_ilan[,1],xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9, main="inverse language")

plot(m_bl_ados[,1],m_gu_ados[,1],xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9, main="ADOS")

```

## Gene ontology biological process

```{r,gobp}

par(mar=c(4,25,1,1))

mgo_gu_ados <- mitch_calc(x=m_gu_ados, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_gu_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_gu_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Guthrie ADOS up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Guthrie ADOS dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Guthrie ADOS GOBP",xlab="S dist")

mgo_gu_diag <- mitch_calc(x=m_gu_diag, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_gu_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_gu_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Guthrie diagnosis up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Guthrie diagnosis dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Guthrie diagnosis GOBP",xlab="S dist")

mgo_gu_iiq <- mitch_calc(x=m_gu_iiq, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_gu_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_gu_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Guthrie inverse IQ up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Guthrie inverse IQ dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Guthrie inverse IQ GOBP",xlab="S dist")

mgo_gu_ilan <- mitch_calc(x=m_gu_ilan, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_gu_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_gu_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Guthrie inverse language up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Guthrie inverse language dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Guthrie inverse language GOBP",xlab="S dist")

mgo_gu_mot <- mitch_calc(x=m_gu_mot, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_gu_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_gu_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Guthrie motor up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Guthrie motor dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Guthrie motor GOBP",xlab="S dist")

mgo_bl_ados <- mitch_calc(x=m_bl_ados, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_bl_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_bl_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Blood at assessment ADOS up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Blood at assessment ADOS dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Blood at assessment ADOS GOBP",xlab="S dist",cex.main=0.75)

mgo_bl_diag <- mitch_calc(x=m_bl_diag, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_bl_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_bl_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Blood at assessment diagnosis up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Blood at assessment diagnosis dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
if ( length(b)>1 ) {
  barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Blood at assessment diagnosis GOBP",xlab="S dist",cex.main=0.75)
}

mgo_bl_iiq <- mitch_calc(x=m_bl_iiq, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_bl_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_bl_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Blood at assessment inverse IQ up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Blood at assessment inverse IQ dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Blood at assessment inverse IQ GOBP",xlab="S dist",cex.main=0.75)

mgo_bl_ilan <- mitch_calc(x=m_bl_ilan, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_bl_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_bl_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Blood at assessment inverse language up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Blood at assessment inverse language dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Blood at assessment inverse language GOBP",xlab="S dist",cex.main=0.7)

mgo_bl_mot <- mitch_calc(x=m_bl_mot, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_bl_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_bl_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Blood at assessment motor up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Blood at assessment motor dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
if ( length(b)>1 ) {
  barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Blood at assessment motor GOBP",xlab="S dist",cex.main=0.7)
}

mgo_buc_ados <- mitch_calc(x=m_buc_ados, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_buc_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_buc_ados$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Buccal ADOS up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Buccal ADOS dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Buccal ADOS GOBP",xlab="S dist")

mgo_buc_diag <- mitch_calc(x=m_buc_diag, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_buc_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_buc_diag$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Buccal diagnosis up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Buccal diagnosis dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Buccal diagnosis GOBP",xlab="S dist")

mgo_buc_iiq <- mitch_calc(x=m_buc_iiq, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_buc_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_buc_iiq$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Buccal inverse IQ up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Buccal inverse IQ dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Buccal inverse IQ GOBP",xlab="S dist")

mgo_buc_ilan <- mitch_calc(x=m_buc_ilan, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_buc_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_buc_ilan$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Buccal inverse language up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Buccal inverse language dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Buccal inverse language GOBP",xlab="S dist")

mgo_buc_mot <- mitch_calc(x=m_buc_mot, genesets=gobp, minsetsize=5, priority="effect",cores=4)
up <- head(subset(mgo_buc_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),20)
dn <- head(subset(mgo_buc_mot$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),20)
up %>% kbl(caption="Buccal motor up") %>% kable_paper("hover", full_width = F)
dn %>% kbl(caption="Buccal motor dn") %>% kable_paper("hover", full_width = F)
b <- c(up$s.dist,dn$s.dist)
names(b) <- c(up$set,dn$set)
b <- b[order(b)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(b))))
names(b) <- substr(names(b), 0, 65)
barplot(abs(b),col=cols,horiz=TRUE,las=1,cex.names=0.7,main="Buccal motor GOBP",xlab="S dist")

```

Multimitch

```{r,multimitchgo}

par( mar = c(5.1, 4.1, 4.1, 2.1) )

gu_l <- list("ADOS"=gu_ados,"diagnosis"=gu_diag,"iIQ"=gu_iiq,"iLang"=gu_ilan,"motor"=gu_mot)
mm_gu <- mitch_import(x=gu_l, DEtype="limma", geneTable=gt )
mres_gu <- mitch_calc(x=mm_gu, genesets=gobp, minsetsize=5, priority="effect",cores=4)
top <- head(subset(mres_gu$enrichment_result,p.adjustMANOVA<0.05),40)
top <- top[,c(1,4:8)]
rownames(top) <- top[,1]
top[,1]=NULL
cols <- colorRampPalette(c("blue", "white", "red"))(n = 25)
colnames(top) <- sub("s.","",colnames(top))
heatmap.2(as.matrix(top),margin=c(6, 24), trace="none",scale="none",col=cols,cexRow=0.7,cexCol=1)
mtext("Guthrie card multicontrast")
if (! file.exists("mres_gu_go.html") ) { mitch_report(res=mres_gu,outfile="mres_gu_go.html") }

bl_l <- list("ADOS"=bl_ados,"diagnosis"=bl_diag,"iIQ"=bl_iiq,"iLang"=bl_ilan,"motor"=bl_mot)
mm_bl <- mitch_import(x=bl_l, DEtype="limma", geneTable=gt )
mres_bl <- mitch_calc(x=mm_bl, genesets=gobp, minsetsize=5, priority="effect",cores=4)
top <- head(subset(mres_bl$enrichment_result,p.adjustMANOVA<0.05),40)
top <- top[,c(1,4:8)]
rownames(top) <- top[,1]
top[,1]=NULL
cols <- colorRampPalette(c("blue", "white", "red"))(n = 25)
colnames(top) <- sub("s.","",colnames(top))
heatmap.2(as.matrix(top),margin=c(6, 24), trace="none",scale="none",col=cols,cexRow=0.7,cexCol=1)
mtext("Blood at assessment multicontrast")
if (! file.exists("mres_bl_go.html") ) { mitch_report(res=mres_bl,outfile="mres_bl_go.html") }

buc_l <- list("ADOS"=buc_ados,"diagnosis"=buc_diag,"iIQ"=buc_iiq,"iLang"=buc_ilan,"motor"=buc_mot)
mm_buc <- mitch_import(x=buc_l, DEtype="limma", geneTable=gt )
mres_buc <- mitch_calc(x=mm_buc, genesets=gobp, minsetsize=5, priority="effect",cores=4)
top <- head(subset(mres_buc$enrichment_result,p.adjustMANOVA<0.05),40)
top <- top[,c(1,4:8)]
rownames(top) <- top[,1]
top[,1]=NULL
cols <- colorRampPalette(c("blue", "white", "red"))(n = 25)
colnames(top) <- sub("s.","",colnames(top))
heatmap.2(as.matrix(top),margin=c(6, 24), trace="none",scale="none",col=cols,cexRow=0.7,cexCol=1)
mtext("Buccal multicontrast")
if (! file.exists("mres_buc_go.html") ) { mitch_report(res=mres_buc,outfile="mres_buc_go.html") }

```

In Guthrie card and blood at assessment I think the genome is undergoing a hypomethylation event,
and some pathways are protected while others experience loss of methylation.
The pathways that appear to have higher methylation might be an artifact of the competitive test.
Methylation of those genes might be relatively stable, but compared to all other genes, it is
relatively higher.
The pathways with loss of methylation include the ones we have already seen, such as taste receptors.
But there are other new ones like glucuronidation, which might be linked to ceellular detoxificaiton.
There are also some inflammatory pathways appearing.

## Session information

For reproducibility

```{r,session}

sessionInfo()

```
