---
title: "Correlates of ADOS and methylation - enrichment analysis"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/asd_meth

## Introduction

Here I will be running a comparison of differential methylation data from guthrie and fresh blood samples.

Here are the files that I'm using:

* limma_blood_ADOS.csv

* limma_guthrie_ADOS.csv

* limma_buccal_ADOS.csv

```{r,libs}

suppressPackageStartupMessages({
  library("limma")
  library("parallel")
  library("mitch")
  library("HGNChelper")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  source("meth_functions.R")
  library("data.table")
  library("kableExtra")
  library("eulerr")
  library("GenomicRanges")
#  data("dualmap850kEID")
  library("RhpcBLASctl")
})

source("meth_functions.R")

RhpcBLASctl::blas_set_num_threads(1)

```

## Annotation

```{r,anno1}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","UCSC_RefGene_Group","Islands_Name","Relation_to_Island")])
head(myann)

gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
gp2 <- lapply(gp2,unique)
gt <- stack(gp2)
colnames(gt) <- c("gene","probe")
gt$probe <- as.character(gt$probe)
dim(gt)
head(gt)

#update gene symbols
new.hgnc.table <- readRDS("new.hgnc.table.rds")
fix <- checkGeneSymbols(gt$gene,map=new.hgnc.table)
fix2 <- fix[which(fix$x != fix$Suggested.Symbol),]
length(unique(fix2$x))
gt$gene <- fix$Suggested.Symbol

```


## Load data

```{r,load1}

bl_mvals <- readRDS(file="bl_mvals.rds")
gu_mvals <- readRDS(file="gu_mvals.rds")
bu_mvals <- readRDS(file="buccal_mvals.rds")

bl_design <- readRDS(file="bl_design_ados.rds")
gu_design <- readRDS(file="gu_design_ados.rds")
bu_design <- readRDS(file="buccal_design_ados.rds")

bl_lim <- read.csv("limma_blood_ADOS.csv")
gu_lim <- read.csv("limma_guthrie_ADOS.csv")
bu_lim <- read.csv("limma_buccal_ADOS.csv")

```

## Mitch import

For each gene, calculate the mean probe t-statistic.

```{r,mitchimport}

blm <- mitch_import(bl_lim,DEtype="limma",geneTable=gt,geneIDcol="Name")

gum <- mitch_import(gu_lim,DEtype="limma",geneTable=gt,geneIDcol="Name")

bum <- mitch_import(bu_lim,DEtype="limma",geneTable=gt,geneIDcol="Name")

```

## Probe level t-stats

```{r,probelevel}

dim(bl_lim)

par(mfrow=c(2,1))

hist(bl_lim$t,xlab="probe t-stat",main="blood at assessment", breaks=30, xlim=c(-4,4))
grid() ; abline(v=0,lty=1,lwd=3)
hist(blm$x,xlab="gene mean t-stat",main="blood at assessment", breaks=30, xlim=c(-2,2))
grid() ; abline(v=0,lty=1,lwd=3)

hist(gu_lim$t,xlab="probe t-stat",main="neonatal Guthrie card", breaks=30, xlim=c(-4,4))
grid() ; abline(v=0,lty=1,lwd=3)
hist(gum$x,xlab="gene mean t-stat",main="neonatal Guthrie card", breaks=30, xlim=c(-2,2))
grid() ; abline(v=0,lty=1,lwd=3)

hist(bu_lim$t,xlab="probe t-stat",main="buccal swab", breaks=30, xlim=c(-4,4))
grid() ; abline(v=0,lty=1,lwd=3)
hist(bum$x,xlab="gene mean t-stat",main="buccal swab", breaks=30, xlim=c(-2,2))
grid() ; abline(v=0,lty=1,lwd=3)

pdf("hist_bl_ados.pdf")
par(mfrow=c(2,1))
hist(bl_lim$t,xlab="probe t-stat",main="blood at assessment", breaks=30, xlim=c(-4,4))
grid() ; abline(v=0,lty=1,lwd=3)
hist(blm$x,xlab="gene mean t-stat",main="blood at assessment", breaks=30, xlim=c(-2,2))
grid() ; abline(v=0,lty=1,lwd=3)
dev.off()

pdf("hist_gu_ados.pdf")
hist(gu_lim$t,xlab="probe t-stat",main="neonatal Guthrie card", breaks=30, xlim=c(-4,4))
grid() ; abline(v=0,lty=1,lwd=3)
hist(gum$x,xlab="gene mean t-stat",main="neonatal Guthrie card", breaks=30, xlim=c(-2,2))
grid() ; abline(v=0,lty=1,lwd=3)
dev.off()

par(mfrow=c(1,1))

```

## Comparison of blood and guthrie card

Now compare gene methylation scores for blood and guthrie cards.
The mean limma t-statistic is used for a scatterplot, then rank-rank plot.

```{r,compare1}

m1 <- merge(blm,gum,by=0)
rownames(m1) <- m1$Row.names
m1$Row.names=NULL
colnames(m1) <- c("bl","gu")

plot(m1$bl,m1$gu,xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9)
grid()
abline(v=0, h=0, lty=2, lwd=2, col="blue")
mtext("aggregated t-statistic for each gene (mean)")

rm1 <- apply(m1,2,rank)
mydiff <- apply(m1,2,function(x) { length(which(x<0)) } )
rm2 <- rm1
rm2[,1] <- rm2[,1] - mydiff[1]
rm2[,2] <- rm2[,2] - mydiff[2]
rnk <- rm2

palette <- colorRampPalette(c("white", "yellow", "orange", "red","darkred", "black"))
xmin = min(rnk[, 1])
xmax = max(rnk[, 1])
ymin = min(rnk[, 2])
ymax = max(rnk[, 2])
k <- MASS::kde2d(rnk[, 1], rnk[, 2])
X_AXIS = "blood at assessment"
Y_AXIS = "neonatal Guthrie card"

filled.contour(k, xlim = c(xmin, xmax), ylim = c(ymin, ymax),
  color.palette = palette, plot.title = {
  abline(v = 0, h = 0, lty = 2, lwd = 2, col = "blue")
  title(main="gene rank",xlab = X_AXIS, ylab = Y_AXIS)
})

pdf("genedens_ados.pdf")
par(mfrow=c(2,2))
plot(1)
plot(1)

plot(m1$bl,m1$gu,xlab="blood at assessment",ylab="neonatal Guthrie card",
  col = rgb(red = 0.6, green = 0.6, blue = 0.6, alpha = 0.5) ,
  pch=19,cex=0.9)
grid()
abline(v=0, h=0, lty=2, lwd=2, col="blue")
mtext("aggregated t-statistic for each gene (mean)")

filled.contour(k, xlim = c(xmin, xmax), ylim = c(ymin, ymax),
  color.palette = palette, plot.title = {
  abline(v = 0, h = 0, lty = 2, lwd = 2, col = "blue")
  title(main = "gene rank", xlab = X_AXIS, ylab = Y_AXIS)
})

dev.off()

```

## Gene set level enrichment

```{r,mitchenrich1}

par(mar = c(5.1, 26.1, 4.1, 2.1))

msigdb <- gmt_import("msigdb.v2023.2.Hs.symbols.gmt")

gobp <- msigdb[grep("^GOBP",names(msigdb))]
gomf <- msigdb[grep("^GOMF",names(msigdb))]
gocc <- msigdb[grep("^GOCC",names(msigdb))]
go <- c(gobp,gomf,gocc)
names(go) <- gsub("_"," ",names(go))

reactome <- gmt_import("ReactomePathways_2024-09-18.gmt")

sfari <- readLines("sfari.txt")
lsfari <- list(sfari)
names(lsfari) <- "sfari"

cores = 8

## BLOOD
# go
bl_mres1 <- mitch_calc(blm,go,priority="effect",cores=cores,minsetsize=5)
bl_mres1tbl <- subset(bl_mres1$enrichment_result,p.adjustANOVA<0.01)
top <- head(bl_mres1tbl,30)

head(top[order(-abs(top$s.dist)),],30) %>%
  kbl(caption="BLOOD: Top effect size GOs found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

# Barplot
up <- head(subset(bl_mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist>0),15)
vup <- up$s.dist
names(vup) <- up$set
dn <- head(subset(bl_mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist<0),15)
vdn <- dn$s.dist
names(vdn) <- dn$set
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.8,xlab="ES",main="Blood at assessment")
mtext("GO")

pdf("ados_barplot_bl_go.pdf")
par(mar = c(5.1, 29.1, 4.1, 2.1))
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.9,xlab="ES",main="Blood at assessment")
mtext("GO")
dev.off()


# reactome
bl_mres2 <- mitch_calc(blm,reactome,priority="effect",cores=cores,minsetsize=5)
bl_mres2tbl <- subset(bl_mres2$enrichment_result,p.adjustANOVA<0.01)
top <- head(bl_mres2tbl,30)

head(top[order(-abs(top$s.dist)),],30) %>%
  kbl(caption="BLOOD: Top effect size Reactome pathways found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

# Barplot
up <- head(subset(bl_mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist>0),15)
vup <- up$s.dist
names(vup) <- up$set
dn <- head(subset(bl_mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist<0),15)
vdn <- dn$s.dist
names(vdn) <- dn$set
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.9,xlab="ES",main="Blood at assessment")
mtext("Reactome")

pdf("ados_barplot_bl_reactome.pdf")
par(mar = c(5.1, 29.1, 4.1, 2.1))
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.9,xlab="ES",main="Blood at assessment")
mtext("Reactome")
dev.off()

## GUTHRIE
# GO
gu_mres1 <- mitch_calc(gum,go,priority="effect",cores=cores,minsetsize=5)
gu_mres1tbl <- subset(gu_mres1$enrichment_result,p.adjustANOVA<0.01)
top <- head(gu_mres1tbl,30)

head(top[order(-abs(top$s.dist)),],30) %>%
  kbl(caption="GUTHRIE: Top effect size GOs found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

# Barplot
up <- head(subset(gu_mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist>0),15)
vup <- up$s.dist
names(vup) <- up$set
dn <- head(subset(gu_mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist<0),15)
vdn <- dn$s.dist
names(vdn) <- dn$set
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.8,xlab="ES",main="Neonatal Guthrie card")
mtext("GO")

pdf("ados_barplot_gu_go.pdf")
par(mar = c(5.1, 29.1, 4.1, 2.1))
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.7,xlab="ES",main="Neonatal Guthrie card")
mtext("GO")
dev.off()

# reactome
gu_mres2 <- mitch_calc(gum,reactome,priority="effect",cores=cores,minsetsize=5)
gu_mres2tbl <- subset(gu_mres2$enrichment_result,p.adjustANOVA<0.01)
top <- head(gu_mres2tbl,30)

head(top[order(-abs(top$s.dist)),],30) %>%
  kbl(caption="GUTHRIE: Top effect size Reactome pathways found with mitch after 1% FDR filtering") %>%
  kable_paper("hover", full_width = F)

# Barplot
up <- head(subset(gu_mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist>0),15)
vup <- up$s.dist
names(vup) <- up$set
dn <- head(subset(gu_mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist<0),15)
vdn <- dn$s.dist
names(vdn) <- dn$set
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.9,xlab="ES",main="Neonatal Guthrie card")
mtext("Reactome")

pdf("ados_barplot_gu_reactome.pdf")
par(mar = c(5.1, 29.1, 4.1, 2.1))
barplot(sort(c(vup,vdn)),horiz=TRUE,las=1,cex.names=0.75,xlab="ES",main="Neonatal Guthrie card")
mtext("Reactome")
dev.off()

par(mar = c(5.1, 4.1, 4.1, 2.1))

```

Mitch joint analysis of blood and guthrie card.
As using the median value is apparently the most sensitive, we'll go with that.

```{r,mitch_joint}

m <- merge(blm,gum,by=0)
rownames(m) <- m$Row.names
m$Row.names = NULL
colnames(m) <- c("bl","gu")
m[is.na(m)] <- 0
res1 <- mitch_calc(m, go, priority="effect",cores=cores)
mitch_report(res1,"blgu_mitch_go.html",overwrite=TRUE)
mitch_plots(res=res1,outfile="blgu_mitch_go_ados.pdf")
sig1 <- subset(res1$enrichment_result,p.adjustMANOVA<0.01)

head(sig1[order(-abs(sig1$s.dist)),],30) %>%
  kbl(caption="Top GOs found with mitch (effect size after 1% FDR filtering)") %>%
  kable_paper("hover", full_width = F)

res2 <- mitch_calc(m, reactome, priority="effect",cores=cores)
mitch_report(res2,"blgu_mitch_reactome.html",overwrite=TRUE)
mitch_plots(res=res2,outfile="blgu_mitch_reactome_ados.pdf")
sig2 <- subset(res2$enrichment_result,p.adjustMANOVA<0.01)

head(sig2[order(-abs(sig2$s.dist)),],30) %>%
  kbl(caption="Top Reactomes found with mitch (effect size after 1% FDR filtering)") %>%
  kable_paper("hover", full_width = F)

```

Top table of genes

```{r,toptable1}

rnk <- res1$ranked_profile

tbl <- data.frame(table(gt$gene))

toptab <- cbind(m,rnk)

toptab$Row.names = NULL

toptab$med <- apply(data.frame(toptab$bl,toptab$gu),1,median)

toptab2 <- merge(toptab,tbl,by.x=0,by.y="Var1")

toptab2 <- toptab2[order(toptab2$med),]

toptab2 <- subset(toptab2,Freq>=5)

head(toptab2,100) %>%
  kbl(caption="Top genes with reducing methylation") %>%
  kable_paper("hover", full_width = F)

```

Now look at the pathways with hypomethylation.

```{r,hypo_go}

hypo <- subset(res1$enrichment_result,p.adjustMANOVA<0.01 & s.bl <0 & s.gu < 0 )
nrow(hypo)

hypo <- hypo[nrow(hypo):1,]
hypo <- tail(hypo,15)

par(mar=c(9.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set,
  col=gray.colors(2), cex.names=0.8,
  xlab="enrichment score")

legend("bottomright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"),
  fill=gray.colors(2), horiz=FALSE, cex=1)

pdf("adosbarplot_hypo_go.pdf")

par(mar=c(4.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set, 
  col=gray.colors(2),
  xlab="enrichment score")

legend("topright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"), 
  fill=gray.colors(2), horiz=FALSE, cex=1)

dev.off()


mar = c(5.1, 4.1, 4.1, 2.1)

```

Now look at bitter taste receptors.

```{r,bitter1}

genesets <- go

bitter <- genesets[grep("GOMF BITTER",names(genesets))]
bitter
rnk2 <- rnk[which(rownames(rnk) %in% bitter[[1]]),]

plot(rnk2,main=names(bitter),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

pdf("bitter.pdf")

plot(rnk2,main=names(bitter),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

dev.off()

```

Heatmap

```{r,bitter2}

colcolfunc <- colorRampPalette(c("yellow", "orange"))

colcols <- colcolfunc(20)[bl_design[,"ADOS"]]

bitter_probes <- gt[which(gt$gene %in% bitter[[1]]),"probe"]

bitter_mvals <- bl_mvals[rownames(bl_mvals) %in% bitter_probes,]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(bitter_mvals,trace="none",col=colfunc(25), margins = c(5,8), scale="row",
  main="Bitter taste receptors",
  ColSideColors = colcols )

```

## Now with reactome

```{r,hypo_reactome}

hypo <- subset(res2$enrichment_result,p.adjustMANOVA<0.01 & s.bl <0 & s.gu < 0 )
nrow(hypo)

hypo <- hypo[nrow(hypo):1,]
hypo <- tail(hypo,15)

par(mar=c(4.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set,
  col=gray.colors(2),
  xlab="enrichment score")

legend("topright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"),
  fill=gray.colors(2), horiz=FALSE, cex=1)

pdf("adosbarplot_hypo_reactome.pdf")

par(mar=c(4.1, 25.1, 5.1, 1.1))

barplot(t(abs(hypo[,c("s.bl","s.gu")])),
  beside=TRUE,horiz=TRUE,las=1,names.arg=hypo$set,
  col=gray.colors(2),
  xlab="enrichment score")

legend("topright", inset=0, title="tissue",
  c("blood at assessment","neonatal Guthrie card"),
  fill=gray.colors(2), horiz=FALSE, cex=1)

dev.off()

mar = c(5.1, 4.1, 4.1, 2.1)

```

Now look at metatropic glutamate/pheromone receptors.

```{r,phero}

genesets <- reactome

phero <- genesets[grep("pheromone",names(genesets))]
phero
rnk2 <- rnk[which(rownames(rnk) %in% phero[[1]]),]

plot(rnk2,main=names(phero),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

pdf("phero.pdf")

plot(rnk2,main=names(phero),xlim=c(-21569,4146),ylim=c(-17678,8037),
  xlab="Rank in blood at assessment",
  ylab="Rank in neonatal Guthrie card",
  cex=1.5,pch=19,col="lightgray")

text(rnk2,labels=rownames(rnk2))

abline(v=0,h=0,lwd=2,lty=2)

dev.off()

```

Make a heatmap of top TAS2R gene probes.

```{r,tar2r}

colcolfunc <- colorRampPalette(c("yellow", "orange"))

colcols <- colcolfunc(20)[bl_design[,"ADOS"]]

phero_probes <- gt[which(gt$gene %in% phero[[1]]),"probe"]

phero_mvals <- bl_mvals[rownames(bl_mvals) %in% phero_probes,]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(phero_mvals,trace="none",col=colfunc(25), margins = c(5,8), scale="row",
  main="Class C/3 (Metabotropic glutamate/pheromone receptors) probes",
  ColSideColors = colcols )

```

## SFARI

Now let's take a look at the SFARI gene list.

```{r,mitch_SFARI}

str(lsfari)

sfarires <- mitch_calc(m1, lsfari, priority="effect")
t(sfarires$enrichment_result)

lrand <-lapply(1:1034,function(i) { set.seed(i*1000) ; sample(rownames(m1),1000) })
names(lrand) <- paste("random",1:1034)
lsets <- c(lsfari,lrand)

sfarires2 <- mitch_calc(m1, lsets, priority="effect",cores=1,resrows=10)

head(sfarires2$enrichment_result) %>%
  kbl(caption = "Enrichment analysis of SFARI genes") %>%
  kable_paper("hover", full_width = F)

mitch_plots(res=sfarires2,outfile="sfari_mitch_ados.pdf")

mitch_report(res=sfarires2,outfile="sfari_mitch_ados.html",overwrite=TRUE)

```

## Buccal pathway analysis

GO

```{r,buccalgo}

bumg <- mitch_calc(bum, go, priority="effect",cores=8)

bumgres <- bumg$enrichment_result

head(subset(bumgres,s.dist<0 ),10) %>%
  kbl(caption="Buccal: Top gene sets with reducing methylation") %>%
  kable_paper("hover", full_width = F)

head(subset(bumgres,s.dist>0 ),10) %>%
  kbl(caption="Buccal: Top gene sets with higher methylation") %>%
  kable_paper("hover", full_width = F)

mysets <- c("GOMF BITTER TASTE RECEPTOR ACTIVITY",
"GOMF TYPE I INTERFERON RECEPTOR BINDING",
"GOBP CELLULAR GLUCURONIDATION",
"GOBP POSITIVE REGULATION OF PEPTIDYL SERINE PHOSPHORYLATION OF STAT PROTEIN",
"GOMF ODORANT BINDING",
"GOMF OLFACTORY RECEPTOR ACTIVITY",
"GOBP URONIC ACID METABOLIC PROCESS")

bumgres[which(bumgres$set %in% mysets),] %>%
  kbl(caption="Selected gene sets in buccal samples") %>%
  kable_paper("hover", full_width = F)

MAX=max(bumg$ranked_profile)
MIN=min(bumg$ranked_profile)

par(mfrow=c(3,1))
par(mar=c(4.1, 4.1, 5.1, 1.1))

beeswarm(bumg$detailed_sets["GOMF BITTER TASTE RECEPTOR ACTIVITY"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("GOMF BITTER TASTE RECEPTOR ACTIVITY (FDR=4.8e-8)",cex=0.7)

beeswarm(bumg$detailed_sets["GOMF TYPE I INTERFERON RECEPTOR BINDING"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("GOMF TYPE I INTERFERON RECEPTOR BINDING (FDR=5.1e-6)",cex=0.7)

mygenes <- go[[which(names(go)=="GOBP CELLULAR GLUCURONIDATION")]]
myranks <- bumg$ranked_profile[which(rownames(bumg$ranked_profile) %in% mygenes),]

beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("GOBP CELLULAR GLUCURONIDATION (FDR=0.24)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)

pdf("buccal_go.pdf")
par(mfrow=c(3,1))
par(mar=c(4.1, 4.1, 5.1, 1.1))

beeswarm(bumg$detailed_sets["GOMF BITTER TASTE RECEPTOR ACTIVITY"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("GOMF BITTER TASTE RECEPTOR ACTIVITY (FDR=4.8e-8)",cex=0.7)

beeswarm(bumg$detailed_sets["GOMF TYPE I INTERFERON RECEPTOR BINDING"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("GOMF TYPE I INTERFERON RECEPTOR BINDING (FDR=5.1e-6)",cex=0.7)

mygenes <- go[[which(names(go)=="GOBP CELLULAR GLUCURONIDATION")]]
myranks <- bumg$ranked_profile[which(rownames(bumg$ranked_profile) %in% mygenes),]

beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("GOBP CELLULAR GLUCURONIDATION (FDR=0.24)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)
dev.off()

```

REACTOME

```{r,buccalreactome}

bumr <- mitch_calc(bum, reactome, priority="effect",cores=8)

bumrres <- bumr$enrichment_result

head(subset(bumrres,s.dist<0 ),10) %>%
  kbl(caption="Buccal: Top gene sets with reducing methylation") %>%
  kable_paper("hover", full_width = F)

head(subset(bumrres,s.dist>0 ),10) %>%
  kbl(caption="Buccal: Top gene sets with higher methylation") %>%
  kable_paper("hover", full_width = F)

mysets <- c("Class C/3 (Metabotropic glutamate/pheromone receptors)",
"Scavenging of heme from plasma",
"Sensory perception of sweet, bitter, and umami (glutamate) taste",
"Glucuronidation",
"Sensory perception of taste",
"Digestion",
"Aspirin ADME",
"Expression and translocation of olfactory receptors",
"Olfactory Signaling Pathway",
"Dectin-2 family")

bumrres[which(bumrres$set %in% mysets),] %>%
  kbl(caption="Selected gene sets in buccal samples") %>%
  kable_paper("hover", full_width = F)

MAX=max(bumr$ranked_profile)
MIN=min(bumr$ranked_profile)

par(mfrow=c(3,1))
par(mar=c(4.1, 4.1, 5.1, 1.1))

beeswarm(bumr$detailed_sets["Class C/3 (Metabotropic glutamate/pheromone receptors)"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("Class C/3 (Metabotropic glutamate/pheromone receptors) (FDR=1.4e-5)",cex=0.7)

mygenes <- reactome[[which(names(reactome)=="Expression and translocation of olfactory receptors")]]
myranks <- bumr$ranked_profile[which(rownames(bumr$ranked_profile) %in% mygenes),]
beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("Expression and translocation of olfactory receptors (FDR=8.5e-43)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)

mygenes <- reactome[[which(names(reactome)=="Olfactory Signaling Pathway")]]
myranks <- bumr$ranked_profile[which(rownames(bumr$ranked_profile) %in% mygenes),]
beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("Olfactory Signaling Pathway (FDR=5.4e-42)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)

pdf("buccal_reactome.pdf")
par(mfrow=c(3,1))
par(mar=c(4.1, 4.1, 5.1, 1.1))

beeswarm(bumr$detailed_sets["Class C/3 (Metabotropic glutamate/pheromone receptors)"],
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
  abline(v=c(MIN,MAX),lty=2)
mtext("Class C/3 (Metabotropic glutamate/pheromone receptors) (FDR=1.4e-5)",cex=0.7)

mygenes <- reactome[[which(names(reactome)=="Expression and translocation of olfactory receptors")]]
myranks <- bumr$ranked_profile[which(rownames(bumr$ranked_profile) %in% mygenes),]
beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("Expression and translocation of olfactory receptors (FDR=8.5e-43)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)

mygenes <- reactome[[which(names(reactome)=="Olfactory Signaling Pathway")]]
myranks <- bumr$ranked_profile[which(rownames(bumr$ranked_profile) %in% mygenes),]
beeswarm(myranks,
  horizontal=TRUE,labels="", xlim=c(MIN,MAX), bty="n",
  xlab="ranked score in buccal swab")
mtext("Olfactory Signaling Pathway (FDR=5.4e-42)",cex=0.7)
  abline(v=c(MIN,MAX),lty=2)

dev.off()


```

## Session information

For reproducibility

```{r,session}

sessionInfo()

save.image("downstream2_ados.Rdata")

```
